### DISCUSSION TITLE
Support `include` directory for Erlang headers

### DISCUSSION THREAD

**Discussion by maennchen** [gleam-lang/gleam#3816] [Ideas & suggestions]
## Rationale

I would like to add gleam bindings into [`oidcc`](https://github.com/erlef/oidcc). The project is built using Erlang and contains Elixir bindings already. Gleam ones would be a nice addition.

I would like to publish only one package containing all 3 languages if possible instead of creating a separate binding only package.

(Plus that would be the first 3 language package on Hex :smile:)

The erlang code uses records and a few macros defined in `.hrl` header files in the [`include` directory](https://github.com/erlef/oidcc/tree/main/include).

## Issue

When trying to add gleam to the project as well, `gleam build` fails with `can't find include file` errors for the compiled `.erl` files.

See: https://github.com/erlef/oidcc/actions/runs/11768509083/job/32778418446

```
gleam build --warnings-as-errors
  Compiling gleam_stdlib
  Compiling gleeunit
  Compiling jose
  Compiling oidcc
/home/runner/work/oidcc/oidcc/build/dev/erlang/oidcc/_gleam_artefacts/oidcc_provider_configuration_SUITE.erl:11:14: can't find include lib "oidcc/include/oidcc_provider_configuration.hrl"
%   11| -include_lib("oidcc/include/oidcc_provider_configuration.hrl").
%     |              ^

/home/runner/work/oidcc/oidcc/build/dev/erlang/oidcc/_gleam_artefacts/oidcc_jwt_util.erl:5:[10](https://github.com/erlef/oidcc/actions/runs/11768509083/job/32778418446#step:6:11): can't find include file "internal/doc.hrl"
%    5| -include("internal/doc.hrl").
%     |          ^

/home/runner/work/oidcc/oidcc/build/dev/erlang/oidcc/_gleam_artefacts/oidcc_provider_configuration_SUITE.erl:26:[13](https://github.com/erlef/oidcc/actions/runs/11768509083/job/32778418446#step:6:14): record oidcc_provider_configuration undefined
%   26|             #oidcc_provider_configuration{
%     |             ^
```

## Reproduction

* Define a normal rebar3 based erlang project
* Add a record in a `.hrl` file
* Include header / use record in an `.erl` source file
* Add `gleam.toml`
* Run `gleam build`

## Versions

* erlang 27.1.2
* rebar 3.24.0
* elixir 1.17.3
* gleam 1.5.1
* OS: Ubuntu 22.04 LTS (GH Action)

## Logs

`GLEAM_LOG=trace gleam build --warnings-as-errors &> gleam_build.log` =>
[gleam_build.log](https://github.com/user-attachments/files/17693094/gleam_build.log)

**Comment by lpil** [comment]
Hello! Even with this I don't think it would be sufficient to implement what you want. Gleam isn't like Elixir and Erlang in that it distributes and directly uses a large amount of modules, so you'd need to add additional packages for the Gleam version to work. This would mean that you'd have to add Gleam-specific dependencies to your package, which would break compilation for Elixir and Erlang users as their build tools don't yet support Gleam. If they did support Gleam then you'd still have a substantial amount of dead code being compiled and loaded for these languages.

**Reply by maennchen** [reply]
Oh, I didn't see that gleam packages publish .hrl, .erl and .gleam files. I assumed that it will only publish .gleam files and compile it on the users machine.

In that case I agree, I don't want to ship and load gleam compiled .erl modules for erlang / elixir users. (I wouldn't mind the extra untouched `.gleam` source files, but nothing at runtime.)

I hoped to be able to do it like I'm doing it with Erlang / Elixir:

* Erlang code is always compiled / loaded
* Elixir is only compiled if Mix is used

**Reply by lpil** [reply]
You would publish a second package that depends on your core package and also the required Gleam dependencies. This gives all users the optimal code bundle.
